OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs

```diff
@@ -63,17 +63,27 @@ void INotifySelected.Selected(Actor self)
 			ShowTargetLines(self);
 		}
 
-		IEnumerable<IRenderable> IRenderAboveShroud.RenderAboveShroud(Actor self, WorldRenderer wr)
+		bool ShouldRender(Actor self)
 		{
 			if (!self.Owner.IsAlliedWith(self.World.LocalPlayer) || Game.Settings.Game.TargetLines == TargetLinesType.Disabled)
-				yield break;
+				return false;
 
 			// Players want to see the lines when in waypoint mode.
 			var force = Game.GetModifierKeys().HasModifier(Modifiers.Shift) || self.World.OrderGenerator is ForceModifiersOrderGenerator;
 
-			if (Game.RunTime > lifetime && !force)
-				yield break;
+			return force || Game.RunTime <= lifetime;
+		}
+
+		IEnumerable<IRenderable> IRenderAboveShroud.RenderAboveShroud(Actor self, WorldRenderer wr)
+		{
+			if (!ShouldRender(self))
+				return Enumerable.Empty<IRenderable>();
+
+			return RenderAboveShroud(self, wr);
+		}
 
+		IEnumerable<IRenderable> RenderAboveShroud(Actor self, WorldRenderer wr)
```

Comment:
Is there any real benefit in creating a new method, or was this just for readability?

Reply:
This is about the eager and lazy evaluation of the enumerable. See https://stackoverflow.com/questions/30176121/why-does-this-string-extension-method-not-throw-an-exception for some context.

If you have a single method with `yield`s in it, the whole method is lazily evaluated. In the example above, this is a problem because we want to throw an argument exception when you call the method initially, not when you try and enumerate the result. The fix is to split the method into two, the first one allows any eager evaluation to be done, and the remaining lazily-evaluated logic goes into a second method which may use continue to use `yield`.

We're using the same pattern here to avoid allocations. Making a call to a method containing `yield`s must allocate an enumerable. We split things into two. If the eager checks mean we know there's no elements to return, we can use `return Enumerable.Empty<IRenderable>();` to bail early. This means we don't make a call to the method which contains `yield`s and thus no allocation is needed.

```c#
IEnumerable<IRenderable> IRenderAboveShroud.RenderAboveShroud(Actor self, WorldRenderer wr)
{
    // These checks are done eagerly as soon as the method is called (a change in behaviour, they used to be lazy and would be done when trying to enumerate instead)
    if (!ShouldRender(self))
        return Enumerable.Empty<IRenderable>();

    // This line allocates as soon as we call it - even though we haven't attempted to enumerate the results yet.
    return RenderAboveShroud(self, wr);
}
```

This change is safe to make because we don't rely on the difference between the eager and lazy evaluation in our rendering code. If we did rely on it, this would be a breaking change we could not make.

Reply:
Thanks for the explanation!